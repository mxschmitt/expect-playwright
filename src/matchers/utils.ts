import type { Page, ElementHandle, Frame } from "playwright-core"
import { PageWaitForSelectorOptions } from "../../global"

export type ExpectInputType = Page | Frame | ElementHandle

const isElementHandle = (value: ExpectInputType): value is ElementHandle => {
  return value.constructor.name === "ElementHandle"
}

export const getFrame = async (value: ExpectInputType) => {
  return isElementHandle(value) ? value.contentFrame() : value
}

const isObject = (value: unknown) =>
  typeof value === "object" && !(value instanceof RegExp)

export type InputArguments = [
  Page | ElementHandle,
  string?,
  (string | PageWaitForSelectorOptions)?,
  PageWaitForSelectorOptions?
]

export const getElementHandle = async (
  args: InputArguments,
  valueArgCount = 1
) => {
  // Pluck the options off the end first
  const options =
    args.length > 1 && isObject(args[args.length - 1])
      ? (args.pop() as PageWaitForSelectorOptions)
      : {}

  // Next, pluck the number of args required by the matcher (defaults to 1)
  const expectedValue = args.splice(-valueArgCount, valueArgCount) as string[]

  // Finally, we can find the element handle
  let elementHandle = (await getFrame(args[0])) ?? args[0]

  // If the user provided a page or iframe, we need to locate the provided
  // selector or the `body` element if none was provided.
  if (!isElementHandle(elementHandle)) {
    const selector = args[1] ?? "body"

    try {
      elementHandle = (await elementHandle.waitForSelector(selector, options))!
    } catch (err) {
      throw new Error(`Timeout exceed for element ${quote(selector)}`)
    }
  }

  return [elementHandle, expectedValue] as const
}

export const quote = (val: string | null) => (val === null ? "" : `'${val}'`)

export const getMessage = (
  { isNot, promise, utils, expand }: jest.MatcherContext,
  matcher: string,
  expected: unknown,
  received: unknown,
  expectedHint: string | undefined = undefined
) => {
  const message = isNot
    ? `Expected: not ${utils.printExpected(expected)}`
    : utils.printDiffOrStringify(
        expected,
        received,
        "Expected",
        "Received",
        expand
      )

  return (
    utils.matcherHint(matcher, undefined, expectedHint, { isNot, promise }) +
    "\n\n" +
    message
  )
}

export const compareText = (
  expectedValue: string | RegExp,
  actualValue: string | null
) => {
  return typeof expectedValue === "string"
    ? expectedValue === actualValue
    : expectedValue.test(actualValue ?? "")
}
