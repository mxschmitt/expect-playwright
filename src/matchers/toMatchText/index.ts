import { SyncExpectationResult } from "expect/build/types"
import { getElementText, getMessage, InputArguments } from "../utils"

const toMatchText: jest.CustomMatcher = async function (
  ...args: InputArguments
): Promise<SyncExpectationResult> {
  try {
    const { elementHandle, expectedValue } = await getElementText(...args)
    /* istanbul ignore next */
    const actualTextContent = await elementHandle.evaluate(
      (el) => el.textContent
    )
    const regex = expectedValue;
    const res = actualTextContent?.match(regex) || [];

    return {
      pass: res.length > 0,
      message: () =>
        getMessage(this, "toMatchText", expectedValue, actualTextContent),
    }
  } catch (err) {
    return {
      pass: false,
      message: () => err.toString(),
    }
  }
}

export default toMatchText
