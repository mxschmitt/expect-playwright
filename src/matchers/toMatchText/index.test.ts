import toMatchText from "."
import { assertSnapshot } from "../tests/utils"

expect.extend({ toMatchText })

describe("toMatchText", () => {
  afterEach(async () => {
    await page.setContent("")
  })
  describe("selector", () => {
    it("positive frame", async () => {
      await page.setContent(`<iframe src="https://example.com"></iframe>`)
      const iframe = await page.$("iframe")
      await expect(iframe!).toMatchText("h1", /.*ample Domai.*/)
    })
    it("positive", async () => {
      await page.setContent(`<div id="foobar">Bar</div>`)
      await expect(page).toMatchText("#foobar", /[B|b]ar/)
    })
    it("negative", async () => {
      await page.setContent(`<div id="foobar">zzzBarzzz</div>`)
      await assertSnapshot(() => expect(page).toMatchText("#foobar", /[b]ar/))
    })
    describe("with 'not' usage", () => {
      it("positive in frame", async () => {
        await page.setContent(`<iframe src="https://example.com"></iframe>`)
        const iframe = await page.$("iframe")
        await expect(iframe!).not.toMatchText("h1", /ab+c/)
      })
      it("positive", async () => {
        await page.setContent(`<div id="foobar">Bar</div>`)
        await expect(page).not.toMatchText("#foobar", /exam.*[1|2]/)
      })
      it("negative", async () => {
        await page.setContent(`<div id="foobar">zzzBarzzz</div>`)
        await assertSnapshot(() =>
          expect(page).toMatchText("#foobar", /not-existing/)
        )
      })
    })
    describe("timeout", () => {
      it("positive: should be able to use a custom timeout", async () => {
        setTimeout(async () => {
          await page.setContent(`<div id="foobar">Bar</div>`)
        }, 500)
        await expect(page).toMatchText("#foobar", /[B|b]ar/)
      })
      it("should throw an error after the timeout exceeds", async () => {
        const start = new Date().getTime()
        await assertSnapshot(() =>
          expect(page).toMatchText("#foobar", /[B|b]ar/, { timeout: 1 * 1000 })
        )
        const duration = new Date().getTime() - start
        expect(duration).toBeLessThan(1500)
      })
    })
  })
  describe("element", () => {
    it("positive", async () => {
      await page.setContent(`<div id="foobar">Bar</div>`)
      const element = await page.$("#foobar")
      expect(element).not.toBeNull()
      await expect(element!).toMatchText("Bar")
    })
    it("negative", async () => {
      await page.setContent(`<div id="foobar">zzzBarzzz</div>`)
      const element = await page.$("#foobar")
      expect(element).not.toBeNull()
      await assertSnapshot(() => expect(element!).toMatchText("not-existing"))
    })
  })
  describe("page", () => {
    it("positive", async () => {
      await page.setContent(`<body><div>Bar</div></body>`)
      await expect(page).toMatchText("Bar")
    })
    it("negative", async () => {
      await page.setContent(`<body><div>zzzBarzzz</div></body>`)
      await assertSnapshot(() => expect(page).toMatchText("not-existing"))
    })
  })
})
