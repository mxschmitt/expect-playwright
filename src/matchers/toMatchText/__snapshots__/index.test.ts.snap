// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toMatchText element negative 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mtoMatchText[2m([22m[32mexpected[39m[2m)[22m

Expected: [32m\\"not-existing\\"[39m
Received: [31m\\"zzzBarzzz\\"[39m"
`;

exports[`toMatchText page negative 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mtoMatchText[2m([22m[32mexpected[39m[2m)[22m

Expected: [32m\\"not-existing\\"[39m
Received: [31m\\"zzzBarzzz\\"[39m"
`;

exports[`toMatchText selector negative 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mtoMatchText[2m([22m[32mexpected[39m[2m)[22m

Expected: [32m/[b]ar/[39m
Received: [31m\\"zzzBarzzz\\"[39m"
`;

exports[`toMatchText selector timeout should throw an error after the timeout exceeds 1`] = `"Error: Timeout exceed for element '#foobar'"`;

exports[`toMatchText selector with 'not' usage negative 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mtoMatchText[2m([22m[32mexpected[39m[2m)[22m

Expected: [32m/not-existing/[39m
Received: [31m\\"zzzBarzzz\\"[39m"
`;
